components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/HSMBasedInternalCryptoProvider.java

```diff
@@ -0,0 +1,165 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.crypto.provider.hsm;
+
+import iaik.pkcs.pkcs11.Mechanism;
+import iaik.pkcs.pkcs11.Session;
+import iaik.pkcs.pkcs11.objects.Key;
+import iaik.pkcs.pkcs11.objects.PrivateKey;
+import iaik.pkcs.pkcs11.objects.PublicKey;
+import iaik.pkcs.pkcs11.wrapper.PKCS11Constants;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.api.ServerConfigurationService;
+import org.wso2.carbon.crypto.api.CryptoException;
+import org.wso2.carbon.crypto.api.InternalCryptoProvider;
+import org.wso2.carbon.crypto.provider.hsm.cryptoprovider.objecthandlers.KeyHandler;
+import org.wso2.carbon.crypto.provider.hsm.cryptoprovider.operators.Cipher;
+import org.wso2.carbon.crypto.provider.hsm.cryptoprovider.util.MechanismDataHolder;
+import org.wso2.carbon.crypto.provider.hsm.cryptoprovider.util.MechanismResolver;
+import org.wso2.carbon.crypto.provider.hsm.cryptoprovider.util.SessionHandler;
+
+import static org.wso2.carbon.crypto.provider.hsm.cryptoprovider.util.CryptoConstants.DECRYPT_MODE;
+import static org.wso2.carbon.crypto.provider.hsm.cryptoprovider.util.CryptoConstants.ENCRYPT_MODE;
+
+/**
+ * Implementation of {@link InternalCryptoProvider} to provide cryptographic operations using Hardware Security Modules.
+ */
+public class HSMBasedInternalCryptoProvider implements InternalCryptoProvider {
+
+    private static final String HSM_BASED_INTERNAL_PROVIDER_KEY_ALIAS_PATH =
+            "CryptoService.HSMBasedCryptoProviderConfig.InternalProvider.KeyAlias";
+
+    private static Log log = LogFactory.getLog(HSMBasedInternalCryptoProvider.class);
+
+    private String keyAlias;
+    private SessionHandler sessionHandler;
+    private MechanismResolver mechanismResolver;
+    private SlotResolver slotResolver;
+
+    /**
+     * Constructor of HSMBasedInternalCryptoProvider. This is an asymmetric crypto provider which caters
+     * internal cryptographic requirements using underlying HSM.
+     *
+     * @param serverConfigurationService : carbon.xml data as a configuration service.
+     */
+    public HSMBasedInternalCryptoProvider(ServerConfigurationService serverConfigurationService)
+            throws CryptoException {
+
+        this.keyAlias = serverConfigurationService.getFirstProperty(HSM_BASED_INTERNAL_PROVIDER_KEY_ALIAS_PATH);
```

Comment:
Overall you can add more debug logs and improve the traceability.

Reply:
Added debug logs to improve the traceability.